### PR TITLE
Fix atomic operations bug for Min and Max

### PR DESCRIPTION
**Pull request by crtrott**
This will fix atomic operations reported on slack, where atomic_min/max would give the wrong answer for a mix of negative/positive values. That was tracked down to a bug in the PTX for the atomic operations inside Desul, and not caught on our side due to test deficiencies.

### CODE REVIEW

File: tpls/desul/include/desul/atomics/cuda/cuda_cc7_asm_atomic_op.inc_isglobal
Comment (dalg24) [inline tpls/desul/include/desul/atomics/cuda/cuda_cc7_asm_atomic_op.inc_isglobal:1]: Drop `__DESUL_IMPL_CUDA_ASM_ATOMIC_INTEGER_OP` and `__DESUL_IMPL_CUDA_ASM_ATOMIC_UNSIGNED_OP` since they are not used any more.

File: tpls/desul/include/desul/atomics/cuda/cuda_cc7_asm_atomic_op.inc_isglobal [hunk 1]
```diff
@@ -76,13 +76,41 @@ __DESUL_IMPL_CUDA_ASM_ATOMIC_INC(type,asm_type,reg_type) \
...
-__DESUL_IMPL_CUDA_ASM_ATOMIC_UNSIGNED_OP(uint32_t,".u32","r")
+// uint32_t
+__DESUL_IMPL_CUDA_ASM_ATOMIC_ADD(uint32_t,".u32","r")
+__DESUL_IMPL_CUDA_ASM_ATOMIC_SUB(uint32_t,".u32","r")
+__DESUL_IMPL_CUDA_ASM_ATOMIC_MIN(uint32_t,".u32","r")
+__DESUL_IMPL_CUDA_ASM_ATOMIC_MAX(uint32_t,".u32","r")
+__DESUL_IMPL_CUDA_ASM_ATOMIC_INC(uint32_t,".u32","r")
+__DESUL_IMPL_CUDA_ASM_ATOMIC_DEC(uint32_t,".u32","r")
+
+// uint64_t
```
Comment (dalg24) [inline tpls/desul/include/desul/atomics/cuda/cuda_cc7_asm_atomic_op.inc_isglobal:93]: Add a note about "inc" and "dec"

File: tpls/desul/include/desul/atomics/cuda/cuda_cc7_asm_atomic_op.inc_predicate [hunk 2]
```diff
@@ -78,29 +78,42 @@ inline __device__ void device_atomic_dec(type* dest, __DESUL_IMPL_CUDA_ASM_MEMOR
...
-__DESUL_IMPL_CUDA_ASM_ATOMIC_UNSIGNED_OP(uint32_t,".u32","r")
+// uint32_t
+__DESUL_IMPL_CUDA_ASM_ATOMIC_ADD(uint32_t,".u32","r")
+__DESUL_IMPL_CUDA_ASM_ATOMIC_SUB(uint32_t,".u32","r")
+__DESUL_IMPL_CUDA_ASM_ATOMIC_MIN(uint32_t,".u32","r")
+__DESUL_IMPL_CUDA_ASM_ATOMIC_MAX(uint32_t,".u32","r")
+__DESUL_IMPL_CUDA_ASM_ATOMIC_INC(uint32_t,".u32","r")
+__DESUL_IMPL_CUDA_ASM_ATOMIC_DEC(uint32_t,".u32","r")
+
+// uint64_t
```
Comment (dalg24) [inline tpls/desul/include/desul/atomics/cuda/cuda_cc7_asm_atomic_op.inc_predicate:96]: Same comment add a note here too

**Review by dalg24 — COMMENTED** [review]
What about `tpls/desul/include/desul/atomics/cuda/cuda_cc7_asm_atomic_op.inc_{forceglobal,generic}`?

Also please confirm that we don't need to update the atomic_fetch_op headers

**Review by masterleinad — CHANGES_REQUESTED** [review]
Needs a corresponding `desul` pull request.

**Comment by crtrott** [comment]
https://github.com/desul/desul/pull/109

File: tpls/desul/include/desul/atomics/cuda/cuda_cc7_asm_atomic_op.inc_predicate (hunk 2 above)
Comment (crtrott) [reply tpls/desul/include/desul/atomics/cuda/cuda_cc7_asm_atomic_op.inc_predicate:102]: done

File: tpls/desul/include/desul/atomics/cuda/cuda_cc7_asm_atomic_op.inc_isglobal (hunk 1 above)
Comment (crtrott) [reply tpls/desul/include/desul/atomics/cuda/cuda_cc7_asm_atomic_op.inc_isglobal:84]: done

File: tpls/desul/include/desul/atomics/cuda/cuda_cc7_asm_atomic_op.inc_isglobal
Comment (crtrott) [reply tpls/desul/include/desul/atomics/cuda/cuda_cc7_asm_atomic_op.inc_isglobal:1]: done

**Comment by crtrott** [comment]
Updated atomic_fetch ops too.

**Comment by masterleinad** [comment]
There are still CI failures in the 32-bit build:
```
 RUN      ] openmp.atomic_operations_complexdouble
atomic_div failed with type N6Kokkos7complexIdEE
atomic_fetch_div failed with type N6Kokkos7complexIdEE
atomic_div_fetch failed with type N6Kokkos7complexIdEE
atomic_div_fetch did not return updated value with type N6Kokkos7complexIdEE
/__w/kokkos/kokkos/core/unit_test/TestAtomicOperations_complexdouble.hpp:40: Failure
Value of: (update != 0 ? atomic_op_test<DivAtomicTest, T, Kokkos::OpenMP>( old_val, update) : true)
  Actual: false
Expected: true
[  FAILED  ] openmp.atomic_operations_complexdouble (0 ms)
[ RUN      ] openmp.atomic_operations_complexfloat
atomic_div failed with type N6Kokkos7complexIfEE
atomic_fetch_div failed with type N6Kokkos7complexIfEE
atomic_div_fetch failed with type N6Kokkos7complexIfEE
atomic_div_fetch did not return updated value with type N6Kokkos7complexIfEE
/__w/kokkos/kokkos/core/unit_test/TestAtomicOperations_complexfloat.hpp:35: Failure
Value of: (update != 0 ? atomic_op_test<DivAtomicTest, T, Kokkos::OpenMP>( old_val, update) : true)
  Actual: false
Expected: true
[  FAILED  ] openmp.atomic_operations_complexfloat (0 ms)
[ RUN      ] openmp.atomic_operations_double
atomic_div failed with type d
atomic_fetch_div failed with type d
atomic_div_fetch failed with type d
atomic_div_fetch did not return updated value with type d
/__w/kokkos/kokkos/core/unit_test/TestAtomicOperations_double.hpp:25: Failure
Value of: (TestAtomicOperations::AtomicOperationsTestNonIntegralType< double, Kokkos::OpenMP>(i, end - i + start, t))
  Actual: false
Expected: true
[  FAILED  ] openmp.atomic_operations_double (0 ms)
[ RUN      ] openmp.atomic_operations_float
atomic_div failed with type d
atomic_fetch_div failed with type d
atomic_div_fetch failed with type d
atomic_div_fetch did not return updated value with type d
/__w/kokkos/kokkos/core/unit_test/TestAtomicOperations_float.hpp:25: Failure
Value of: (TestAtomicOperations::AtomicOperationsTestNonIntegralType< double, Kokkos::OpenMP>(i, end - i + start, t))
  Actual: false
Expected: true
[  FAILED  ] openmp.atomic_operations_float (0 ms)
```

File: core/unit_test/TestAtomicOperations_complexdouble.hpp [hunk 3]
```diff
@@ -37,9 +37,14 @@ TEST(TEST_CATEGORY, atomic_operations_complexdouble) {
         (atomic_op_test<SubAtomicTest, T, TEST_EXECSPACE>(old_val, update)));
     ASSERT_TRUE(
         (atomic_op_test<MulAtomicTest, T, TEST_EXECSPACE>(old_val, update)));
-    ASSERT_TRUE((update != 0 ? atomic_op_test<DivAtomicTest, T, TEST_EXECSPACE>(
-                                   old_val, update)
-                             : true));
+
+    // disable division test for 32bit where we have accuracy issues with
+    // division atomics still compile it though
```
Comment (masterleinad) [inline core/unit_test/TestAtomicOperations_complexdouble.hpp:42]:
```suggestion
    // FIXME_32BIT disable division test for 32bit where we have accuracy issues with
    // division atomics, still compile it though
```

File: core/unit_test/TestAtomicOperations_complexfloat.hpp [hunk 4]
```diff
@@ -32,9 +32,14 @@ TEST(TEST_CATEGORY, atomic_operations_complexfloat) {
         (atomic_op_test<SubAtomicTest, T, TEST_EXECSPACE>(old_val, update)));
     ASSERT_TRUE(
         (atomic_op_test<MulAtomicTest, T, TEST_EXECSPACE>(old_val, update)));
-    ASSERT_TRUE((update != 0 ? atomic_op_test<DivAtomicTest, T, TEST_EXECSPACE>(
-                                   old_val, update)
-                             : true));
+
+    // disable division test for 32bit where we have accuracy issues with
```
Comment (masterleinad) [inline core/unit_test/TestAtomicOperations_complexfloat.hpp:36]:
```suggestion
    // FIXME_32BIT disable division test for 32bit where we have accuracy issues with
```

File: core/unit_test/TestAtomicOperations_double.hpp [hunk 5]
```diff
@@ -22,8 +22,11 @@ TEST(TEST_CATEGORY, atomic_operations_double) {
   const int end   = 11;
   for (int i = start; i < end; ++i) {
     for (int t = 0; t < 8; t++)
-      ASSERT_TRUE((TestAtomicOperations::AtomicOperationsTestNonIntegralType<
-                   double, TEST_EXECSPACE>(i, end - i + start, t)));
+      // disable division test for 32bit where we have accuracy issues with
```
Comment (masterleinad) [inline core/unit_test/TestAtomicOperations_double.hpp:25]:
```suggestion
      // FIXME_32BIT disable division test for 32bit where we have accuracy issues with
```

File: core/unit_test/TestAtomicOperations_float.hpp [hunk 6]
```diff
@@ -22,8 +22,11 @@ TEST(TEST_CATEGORY, atomic_operations_float) {
   const int end   = 11;
   for (int i = start; i < end; ++i) {
     for (int t = 0; t < 8; t++)
-      ASSERT_TRUE((TestAtomicOperations::AtomicOperationsTestNonIntegralType<
-                   double, TEST_EXECSPACE>(i, end - i + start, t)));
+      // disable division test for 32bit where we have accuracy issues with
```
Comment (masterleinad) [inline core/unit_test/TestAtomicOperations_float.hpp:25]:
```suggestion
      // FIXME_32BIT disable division test for 32bit where we have accuracy issues with
```

**Comment by masterleinad** [comment]
Fine with me if you add some `FIXME_32BIT` comments.

File: core/unit_test/TestAtomicOperations_complexdouble.hpp [hunk 7]
```diff
@@ -16,22 +16,37 @@
...
-                                             TEST_EXECSPACE>(start, end - i)));
+        (atomic_op_test<SubAtomicTest, T, TEST_EXECSPACE>(old_val, update)));
     ASSERT_TRUE(
-        (TestAtomicOperations::DivAtomicTest<Kokkos::complex<double>,
-                                             TEST_EXECSPACE>(start, end - i)));
+        (atomic_op_test<MulAtomicTest, T, TEST_EXECSPACE>(old_val, update)));
+
+    // disable division test for 32bit where we have accuracy issues with
+    // division atomics still compile it though
+    if (sizeof(void*) == 8)
```
Comment (Rombur) [inline core/unit_test/TestAtomicOperations_complexdouble.hpp:43]:
```
Kokkos/core/unit_test/TestAtomicOperations_complexdouble.hpp:43:8: error: suggest explicit braces to avoid ambiguous 'else' [-Werror=dangling-else]

     if (sizeof(void*) == 8)

        ^
```

**Review by masterleinad — APPROVED** [review]

File: tpls/desul/include/desul/atomics/cuda/cuda_cc7_asm_atomic_fetch_op.inc_isglobal [hunk 8]
```diff
@@ -205,4 +222,4 @@ __DESUL_IMPL_CUDA_ASM_ATOMIC_FETCH_BIN_OP()
 #undef __DESUL_IMPL_CUDA_ASM_ATOMIC_FETCH_INC
 #undef __DESUL_IMPL_CUDA_ASM_ATOMIC_FETCH_DEC
 #undef __DESUL_IMPL_CUDA_ASM_ATOMIC_FETCH_AND
-
+#undef __DESUL_IMPL_CUDA_ASM_ATOMIC_FETCH_BIN_OP
```
Comment (ldh4) [inline tpls/desul/include/desul/atomics/cuda/cuda_cc7_asm_atomic_fetch_op.inc_isglobal:225]: Why is this added? Was it just missing before?

**Review by ldh4 — APPROVED** [review]
Mainly looked at changes in TestAtomicOperations files. Those look fine to me.

File: tpls/desul/include/desul/atomics/cuda/cuda_cc7_asm_atomic_fetch_op.inc_isglobal (hunk 8 above)
Comment (crtrott) [reply tpls/desul/include/desul/atomics/cuda/cuda_cc7_asm_atomic_fetch_op.inc_isglobal:225]: yeah
